aquadoggo/src/replication/service.rs

```diff
@@ -0,0 +1,398 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::time::Duration;
+
+use anyhow::Result;
+use libp2p::swarm::ConnectionId;
+use libp2p::PeerId;
+use log::{debug, info, warn};
+use p2panda_rs::schema::SchemaId;
+use tokio::sync::broadcast::Receiver;
+use tokio::task;
+use tokio::time::sleep;
+
+use crate::bus::{ServiceMessage, ServiceSender};
+use crate::context::Context;
+use crate::db::SqlStore;
+use crate::manager::{ServiceReadySender, Shutdown};
+use crate::replication::errors::ReplicationError;
+use crate::replication::{Mode, Session, SyncIngest, SyncManager, SyncMessage, TargetSet};
+use crate::schema::SchemaProvider;
+
+use super::SessionId;
+
+const MAX_SESSIONS_PER_CONNECTION: usize = 3;
+
+#[derive(PartialEq, Eq, PartialOrd, Clone, Debug, Hash)]
+
+/// Identifier for a connection to another peer. The `ConnectionId` is optional as we need to
+/// identify the local peer in the replication manager, but in this case there is no single
+/// connection to associate with.
+///
+/// @TODO: This could be modelled better maybe...
+pub struct PeerConnectionId(PeerId, Option<ConnectionId>);
+
+impl Display for PeerConnectionId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}{:?}",
+            self.0,
+            // If we don't pass a connection id then we default to `0`
+            self.1.unwrap_or(ConnectionId::new_unchecked(0))
+        )
+    }
+}
+
+pub async fn replication_service(
+    context: Context,
+    shutdown: Shutdown,
+    tx: ServiceSender,
+    tx_ready: ServiceReadySender,
+) -> Result<()> {
+    // Subscribe to communication bus
+    let _rx = tx.subscribe();
+
+    let local_peer_id = context
+        .config
+        .network
+        .peer_id
+        .expect("Peer id needs to be given");
+
+    // Define set of schema ids we are interested in
+    let supported_schema_ids: Vec<SchemaId> = context
+        .schema_provider
+        .all()
+        .await
+        .iter()
+        .map(|schema| schema.id().to_owned())
+        .collect();
+    let target_set = TargetSet::new(&supported_schema_ids);
+
+    // Run a connection manager which deals with the replication logic
+    let manager = ConnectionManager::new(
+        &context.schema_provider,
+        &context.store,
+        &tx,
+        local_peer_id,
+        target_set,
+    );
+
+    let scheduler_handle = task::spawn(ConnectionManager::start_scheduler(tx.clone()));
```

Comment:
Wondering if that scheduler can be an async stream which wakes up the runtime every x seconds, there's some nice rust patterns for it. I can look something up!